### DISCUSSION TITLE
WebAssembly

### DISCUSSION THREAD

**Discussion by 4ad** [cue-lang/cue#2045] [Proposal]
This is an umbrella discussion where people can post their thoughts on how they'd hope or expect to use web assembly (Wasm) with CUE.

**Comment by sdboyer** [comment]
We have a solid use case for WASM with Thema - https://github.com/grafana/thema/issues/36.

Basically, we want a Thema-specific CUE playground-ish environment, which lets us present the user with all the basic thema operations. (Essentially, the operations exposed via the `thema` CLI).

Certainly, it would make [docs like this one](https://github.com/grafana/thema/blob/d936c1921950c4517cb9da1cb994340aa9cf6990/docs/go-quickstart.md) far more powerful - instead of asking people to run code generation, try out validation, etc., on their local, they can see it directly in the browser. But it would also enable us to do things like - just one example - update tutorials we write for configuring systems on the fly/in the browser, if their config file is written in Thema.

It's important to note that we need to have some Go logic in there, not just pure CUE, because we can't express all the constraints/invariants required by Thema in native CUE, and because our code generators are written in Go.

**Reply by 4ad** [reply]
If I understand you correctly, you want to run CUE in a WebAssembly host, similar to how the CUE playground currently works. I believe @disintegrator is also looking into something similar.

This is all great, but we are also looking into the complementary use case, where you'd use CUE normally but be able to call into functionality provided as Wasm modules. That is, we're looking into a way to extend what users can do with CUE by being able to define their own functions in their preferred programming language (Go, Rust, etc), and be able to make use of these functions from their CUE configurations.

**Reply by sdboyer** [reply]
Interesting. That's actually a pretty complementary use case for us. Maybe even preferable.

One of the design goals for Thema is providing symmetric APIs in many contexts - CLI, HTTP, Go SDK, native CUE. The WASM use case really just needs _one_ of those, and while the Go SDK is the most obvious and direct path, we plausibly get more by exposing it via native CUE.

Probably a more distant goal for us but the overlap remains notable

**Comment by kajogo777** [comment]
We have a use case for WASM at [DevX GitHub repo](https://github.com/devopzilla/guku-devx)

We want to execute custom pure functions inside CUE manifests (e.g. like list.Concat(...) ) to transform data.

This will allow users of DevX to write custom configuration transformation functions in any language, compile them to WASM then call them from within DevX Transformers.

The transformation functions are pure: data in -> data out, with a fixed input and output schema.

I can also help with the implementation.

**Comment by justenstall** [comment]
There may be a use case for WASM in [Timoni](https://github.com/stefanprodan/timoni) or in CUE itself to support evaluating [CEL expressions](https://github.com/google/cel-spec). Kubernetes uses CEL for validation in CRDs and eventually will use them for their native types, and the `timoni mod vendor crd`/`timoni mod vendor k8s` commmands which translate the schemas to CUE currently ignore these validations. Timoni is looking at running the CEL validations using `k8s.io/apiserver/pkg/apis/cel`, but it would be even better if there was a CUE-native way to do this.

The [CEL Playground project](https://github.com/undistro/cel-playground) produces a WASM binary to run the validations, which could be used in Timoni modules to validate their CUE configs.  

The bigger picture idea is to support translating CEL validations to CUE equivalents when possible, but running them with WASM would be a seriously useful substitute for now.

This is related to #2691.

@stefanprodan
